site/app/models/gradeable/Component.php

```diff
@@ -0,0 +1,215 @@
+<?php
+
+namespace app\models\gradeable;
+
+use app\exceptions\AggregateException;
+use app\exceptions\NotImplementedException;
+use app\libraries\Core;
+use app\models\AbstractModel;
+
+
+/**
+ * Class Component
+ * @package app\models\gradeable
+ *
+ * All data describing the configuration of a Gradeable Component
+ *
+ * @method getId();
+ * @method getTitle();
+ * @method setTitle($title);
+ * @method getTaComment();
+ * @method setTaComment($ta_comment);
+ * @method getStudentComment();
+ * @method setStudentComment($student_comment);
+ * @method getLowerClamp();
+ * @method getDefault();
+ * @method getMaxValue();
+ * @method getUpperClamp();
+ * @method isText();
+ * @method setText($is_text);
+ * @method isPeer();
+ * @method setPeer($is_peer);
+ * @method getOrder();
+ * @method setOrder($order);
+ * @method getPage();
+ * @method setPage($page);
+ */
+class Component extends AbstractModel
+{
+    /** @property @var int The course-wide unique numeric id of this component */
+    protected $id = 0;
+    /** @property @var string The title of this component */
+    protected $title = "";
+    /** @property @var string The comment only visible to the TA/manual grader */
+    protected $ta_comment = "";
+    /** @property @var string The comment visible to the student */
+    protected $student_comment = "";
+    /** @property @var int The minimum points this component can contribute to the score (can be negative) */
+    protected $lower_clamp = 0;
+    /** @property @var int The number of points this component is worth with no marks */
+    protected $default = 0;
+    /** @property @var int The full value of this component without extra credit */
+    protected $max_value = 0;
+    /** @property @var int The maximum number of points this component can contribute to the score (can be > $max_value) */
+    protected $upper_clamp = 0;
+    /** @property @var bool If this is a text component (true) or a numeric component (false) for numeric/text components */
+    protected $text = false;
+    /** @property @var bool If this is a peer grading component */
+    protected $peer = false;
+    /** @property @var int The order of this component in the gradeable */
+    protected $order = -1;
+    /** @property @var int The pdf page this component will reside in */
+    protected $page = -1;
+
+    /** @property @var Mark[] All possible common marks that can be assigned to this component */
+    protected $marks = array();
+
+
```

Comment:
It may be a good idea for components to have a reference back to their Gradeable (so you don't have to pass the entire hierarchy every time). Would be interesting to see if that's doable without making toArray have infinite recursion.

Reply:
Done